### PR TITLE
Resolves #289: Handle processParent in mojos which support it.

### PR DESCRIPTION
**Pull request by ajarmoniuk**
Also:
- added failIfNotReplaced to ForceReleasesMojo as a first steps towards converging with UseReleasesMojo
- added unit tests and integration tests
- some refactoring

More steps towards folding ForceReleases and UseReleases will follow. Also, I see it's possible to reuse parts of the common base class used for lastest versions/releases, but this will follow in future PRs.

### CODE REVIEW

**Comment by ajarmoniuk** [comment]
@slawekjaranowski please review

**Comment by ajarmoniuk** [comment]
I have updated it a bit to fold `UseReleasesMojo` with `ForceReleasesMojo` a bit further, but that's about it in this PR, I'll continue this subject in a separate PR.

Another point to consider: I think #134 is not limited to that one mojo, but may be present everywhere -- to be checked and possibly fixed.

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/reporting/AbstractVersionsReportRenderer.java [hunk 1]
```diff
@@ -64,10 +64,6 @@
     protected ArtifactVersionsCache allUpdatesCache
             = new ArtifactVersionsCache( AbstractVersionDetails::getAllUpdates );
 
-    /**
```
Comment (ajarmoniuk) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/reporting/AbstractVersionsReportRenderer.java:67]: Removed this documentation comment as it was misleading (not used by Plexus) and not really necessary -- the parameters are pretty obvious.

**Comment by ajarmoniuk** [comment]
Sorry about these multiple updates today, I think that's it for this PR.

Now it all looks lean and clean. Almost all mojos updating versions consist of an almost similar method (which by the way could also be folded), and a (small) business method, providing the provider method, instructing `useLatestVersions` how exactly it should retrieve the latest version.

**Comment by ajarmoniuk** [comment]
@slawekjaranowski rebased. Please review.

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java [hunk 2]
```diff
@@ -205,10 +205,11 @@
     @Parameter( property = "changeRecorderOutputFile",
                 defaultValue = "${project.build.directory}/versions-changes.xml" )
     private File changeRecorderOutputFile;
+
     /**
      * The change recorder implementation.
      */
-    private ChangeRecorder changeRecorder;
+    private volatile ChangeRecorder changeRecorder;
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java:212]: Is it accessed from multi thread somehow?

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java [hunk 3]
```diff
@@ -560,24 +574,21 @@ else if ( PomHelper.setPropertyVersion( pom, version.getProfileId(), property.ge
...
+    protected ChangeRecorder getChangeRecorder()
     {
         if ( changeRecorder == null )
         {
-            if ( "none".equals( this.changeRecorderFormat ) )
-            {
-                changeRecorder = ChangeRecorderNull.create();
-            }
-            else if ( "xml".equals( this.changeRecorderFormat ) )
+            synchronized ( this )
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java:582]: why is needed ... streams..parallel ?

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java (hunk 2 above)
Comment (ajarmoniuk) [reply versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java]: Not needed indeed. Removed.

**Comment by slawekjaranowski** [comment]
As I assume - after it PR #289 an issue #288 - can be closed - right?

**Pull request merged**